text/0000-multi-type-return-position-impl-trait.md

```diff
@@ -0,0 +1,673 @@
+
+```rust
+enum Enum<'a> {
+    A(&'a mut std::ops::Range<i32>),
+    B(std::vec::IntoIter<i32>),
+}
+
+impl<'a> Iterator for Enum<'a> {
+    type Item = i32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Enum::A(iter) => iter.next(),
+            Enum::B(iter) => iter.next(),
+        }
+    }
+
+    // ..repeat for the remaining 74 `Iterator` trait methods
+}
+
+fn multi_iter<'a>(x: i32, iter_a: &'a mut std::ops::Range<i32>) -> Enum<'a> {
+    match x {
+        0 => Enum::A(iter_a),
+        _ => Enum::B(vec![5, 10].into_iter()),
+    }
+}
+```
+
+It should be fine if multiple iterators use the same lifetime. But only a single
+lifetime should be permitted on the return type, as is the case today when
+using `dyn Trait`:
+
+```rust
+// ❌ Fails to compile (Rust 1.66.0)
+// error[E0226]: only a single explicit lifetime bound is permitted
+fn fails<'a, 'b>() -> Box<dyn Iterator + 'a + 'b> {
+    ...
+}
+```
+
+# (TODO) Relationship to `dyn`
+
+TODO (explain the relationship to dyn traits, returning `Self`, and which traits
+can and can't be allowed, and how that differs from single-value TAITs).
+
+## (TODO) Parity in lifetime rules
+
+TODO: Move lifetimes section here
+
+## (TODO) Sized and Self
+
+TODO (tldr: neither is an issue with `impl Trait` today, and it shouldn't have
+to be either in the future. That's the main difference in this approach.)
+
+# Drawbacks
+
+## (TODO) Downcasting
+
+TODO (tldr: because we're generating an anonymous enum, you can't downcast to a
+concrete type. With single-value RPIT you have a concrete type which you _can_
+downcast to. But with MTRPIT we're generating a type, meaning you can't downcast
+into it)
+
+## (TODO) Size and performance
+
+TODO (tldr: we already see issues around auto-generated enums in the compiler
+already. Boats did a great post on that. The solution is probably not to eschew
+ergonomics in favor of more "manual" uses, but instead to provide better tools
+to track, diagnose, and mitigate problems as they arise. Both the compiler and
+clippy have knowledge of type sizes. They can, and already do provide hints
+where needed.)
+
+[boats-segmented-stacks]: https://without.boats/blog/futures-and-segmented-stacks/
+[clippy-large-enum]: https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
+
+## (TODO) Teaching `impl` and `dyn`
+
+TODO (tldr: this blurs the line between `impl` and `dyn` somewhat, and we need
+to provide a better explanation than "impl for one type, dyn for many types".
+But that's okay; we're not fundamentally enabling anything new, so the decision
+isn't new either - we just need to better explain the tradeoffs.)
+
+[rbe-dyn]: https://doc.rust-lang.org/stable/rust-by-example/trait/dyn.html
+[trpl-trait-objs]: https://doc.rust-lang.org/book/ch17-02-trait-objects.html
+[trpl-dyn-size]: https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait
+
+# Alternatives
```

Comment:
Is this specific to *return position*?  I agree that `-> enum impl Trait` isn't great, but I think it's not because of syntax, but because the *caller* doesn't actually care that it's an `enum`.  (After all, it could have been an enum -- commonly `Either` -- before anyway.)

Thus I wonder instead about something that opts-into it for the control flow constructs instead, so that it works in places other than return types.  For example, I could imagine resolving the classic "use `Either` for different iterators" problem via something like

```rust
let it = enum if backward { it.rev() } else { it };
for x in it { … }
```

Where the syntax marker is saying "FYI, these enum arms are going to have different types, unlike usual `if`, but that's ok and what I wanted here".  (And there could similarly be `enum match`.)

Reply:
This doesn't extend *super* gracefully to `enum return`, but it works there as well; it just requires all `return` to agree whether it's `enum` or not. Similarly for `enum break`. (Or would it be `enum loop`? Though that wouldn't really work for breaking from labeled blocks.)

(The as of yet unresolved question about whether enum dispatch is just a restriction on what traits can be used or if it has externally visible implications does still remain. If we allow the compiler to flatten anonymous `enum`s to remove nested dispatch[^1], then that's potentially observable via typeid/downcast equivalence in a way not achievable with just local enums.)

[^1]: Which is probably a requisite for it to be "zero cost," as in you couldn't do it any better by hand. Since with non-opaque enums, you could manually flatten the dispatch and get better discriminant niching in addition to the potentially faster resulting enum dispatch table.

Reply:
It's fairly clear `enum Trait` winds up morally more a flavor of `dyn Trait` than a flavor of `impl Trait`, right?   `enum Trait: Trait` is false, just like `dyn Trait: Trait` is false, right?

`dyn Trait` is otoh a type because it's always instantiated the same.  `enum Trait` is not always instantiated the same.  `enum Trait` could be a trait however, roughly like `FnOnce`, `FnMut`, and `Fn`, so then `-> impl enum Trait` makes sense.  Amusingly `dyn Trait: enum Trait`, yes?  In fact `fn() -> Box<impl enum Trait>` could return `Box<dyn Trait>` under the hood?

We produce `dyn Trait`s via casts `as Pointer<dyn Trait>` because this only requite changing pointer metadata.  We do not want this for `enum Trait` so yeah one needs some keyword the constructs the concrete type of the enum, but maybe the compiler could be smarter about this somehow, ala
```
let mut it = enum it;
if backward { it = enum if.rev(); }
```